File: 算法竞赛进阶指南/0x00基本算法/AcWing107.py
N = 500006
n, ans, a, b = 0, 0, [0]*N, [0]*N

def merge(l, mid, r):
    global ans
    if l == r: return
    merge(l, (l+mid)>>1, mid)
    merge(mid+1, (mid+1+r)>>1, r)
    i, j = l, mid+1
    for k in range(l, r+1):
        if j > r or (i <= mid and a[i] <= a[j]): b[k] = a[i]; i += 1
        else: b[k] = a[j]; j += 1; ans += mid - i + 1
    for k in range(l, r+1): a[k] = b[k]

def Ultra_QuickSort():
    global ans
    for i in range(1, n+1): a[i] = int(input())
    ans = 0
    merge(1, (1+n)>>1, n)
    print(ans)
    
def main():
    global n
    while True:
        n = int(input())
        if n == 0: break
        Ultra_QuickSort()

main()

File: 0x00基本算法/0x05排序/AcWing107.cpp
#include <iostream>
#define ll long long
using namespace std;
const int N = 500006;
int n, a[N], b[N];
ll ans;

void merge(int l, int mid, int r) {
  if (l == r) return;
  merge(l, (l+mid)>>1, mid);
  merge(mid+1, (mid+1+r)>>1, r);
  int i = l, j = mid+1;
  // 归并排序左右合并时进行逆序对统计
  for (int k=l; k<=r; k++)
    if (j > r || (i <= mid && a[i] <= a[j])) b[k] = a[i++];
    else b[k] = a[j++], ans += mid - i + 1;
  for (int k=l; k<=r; k++) a[k] = b[k];
}

void Ultra_QuickSort() {
  for (int i=1; i<=n; i++) cin >> a[i];
  ans = 0;
  merge(1, (1+n)>>1, n);
  cout << ans << endl;
}

int main() {
  while (cin >> n && n) Ultra_QuickSort();
  return 0;
}